### PR TITLE
CRAB-33767: Incorporate Snapshots and AMIs in NagBot

### PR DESCRIPTION
**Pull request by alizbay94**
This PR introduces changes that allows NagBot to process AMIs and snapshots when issuing the 'execute' and 'notify' commands. 

**Pricing for Snapshots** was determined using the average price of storing _standard_ Snapshots in us-west-1, us-west-2, us-east-1, and us-east-2. I did not include logic to estimate the price of _archived_ snapshots since Seeq only uses standard snapshots to-date.

**Pricing for AMIs** was determined by taking the sum total price of the snapshots comprising the AMI. Only pricing for 
_EBS-Backed_ AMI's was implemented since Seeq does not use _instance-backed_ AMIs to-date.

**Terminating AMIs [considerations](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/deregister-ami.html#clean-up-ebs-ami)**:
- When you deregister an EBS-backed AMI, it doesn't affect the snapshot(s) that were created for the volume(s) of the instance during the AMI creation process. You'll continue to incur storage costs for the snapshots. I did not implement logic to explicitly delete the snapshots comprising EBS-backed AMIs since ideally those snapshots would have the same terminate date as the AMI upon creation. NagBot processes AMIs before snapshots, so the AMI would be deregistered then the snapshots comprising it would be deleted soon-after if they have the same terminate date. I can still implement the logic to also delete the snapshots when the AMI is deregistered if it sounds desirable, the snapshot ids are privided by the `boto3.resource.Image()` so it should only be a matter of calling `terminate_resource` on those snapshots.
**Note** After further discussion, functionality has been implemented in this PR so that any Snapshots making up an EBS-backed AMI are deleted when NagBot deregisters the AMI. Furthermore, pricing for Snapshots belonging to EBS-Backed AMIs are not included in the monthly Snapshot pricing, but they are instead included in the monthly AMI pricing.

**Terminating Snapshot [considerations](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-deleting-snapshot.html#ebs-delete-snapshot-considerations)**:
- You can't delete a snapshot of the root device of an EBS volume used by a registered AMI. You must first deregister the AMI before you can delete the snapshot. For this reason, AMI's are processed before snapshots when NagBot runs so that the AMI is deregistered before the snapshots comprising it are deleted. I didn't implement logic to deregister the AMI which the snapshot is a part of in this scenario, since the AMI and the snapshots it is composed of should have the same terminate date if we implement an AWS Lambda function in the future that automatically tags resources. Since AMIs are processed by NagBot first, the AMI would be deregistered if its terminate date comes up, then the snapshots would be able to be deleted - the snapshots would be deleted when NagBot processes our organizations snapshots.
- You can't delete a snapshot that is managed by the AWS Backup service using Amazon EC2. Instead, you must use AWS Backup to delete the corresponding recovery points in the backup vault - I did not implement logic for this scenario, but I imagine I could if it is desirable.

**Pytests have been written** for the `Snapshot` and `Ami` classes. 3/5 tests written for each class are identical to the tests written for the `Volume` class, with the exception of 2 tests involving terminating resources. Since AMIs and snapshots are deleted differently with the python `boto3` library than how EBS volumes and EC2 instances are deleted, I set the return value the aws `Snapshot()`and `Image()` classes to a MagicMock() - eventually it is asserted that the MagicMock() called `delete()` and `deregister()` for snapshots and ami respectively during `terminate_resource`.

**`README.md` has been updated** with instructions on how to update NagBot's version. Mentioning of volumes, AMIs, and snapshots has also been included in the README. A new slack app for testing NagBot has been created and added to the #bot-testing channel in slack., so the`README.md` was also updated with how to test NagBot during development, and where to find the `SLACK_BOT_TOKEN` for the NagBotTest slack app on Keeper.

**Some refactoring was done** as part of the changes in this PR so that `format()` strings were converted to f strings for readability.

---

Future Improvements: I noticed that our organization's snapshots and AMIs are often untagged, implementing a method of auto-tagging resources in the future (potentially using an AWS Lambda) would be beneficial to ensure those resources get tagged and deleted when they are no longer needed.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,6 +25,13 @@ Here's what a Nagbot notification looks like in Slack:
 
 Here is a [recording of the process to setup a Slack app](https://seeq.zoom.us/rec/share/qgmqAvz_2eV3SYiNJO4mLrQlH94eGXSs89BDSl28Epl-Bjey9_DgvBLnkF3W2dOf.Zd12Sq-G5QbsOPrt) (Passcode: z?mJ%d08). The process frequently changes due to Slack changing the UI for applications.
 
+# Testing Nagbot changes during development
+
+The NagBotTest app has been registered to slack, and it is set up to send NagBot output to slack. The default 
+testing channel for NagBot is #bot-testing. 
+
+The SLACK_BOT_TOKEN for NagBotTest is found on Keeper [here](https://keepersecurity.com/vault/#detail/AUj3FYXz44uON4CVQSTKMQ).
```
Comment (CodyRay) [inline README.md]: :+1: Remember that this repository is public, so consider prefacing this to be for Seeq employees in case anyone stumbles upon nagbot

File: app/ami.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
+from dataclasses import dataclass
+
+from app import resource
+from .resource import Resource
+import boto3
+
+from datetime import datetime
+TODAY = datetime.today()
```
Comment (CodyRay) [inline app/ami.py:10]: (sanity check) are any changes needed here from #11?

File: app/ami.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        return [self.resource_id,
+                self.name,
+                self.state,
+                self.terminate_after,
+                self.contact,
+                self.monthly_price,
+                self.region_name,
+                self.resource_type,
+                self.operating_system,
+                self.iops,
```
Comment (CodyRay) [inline app/ami.py:53]: Is storage type (gp2, gp3, for example) also available?

File: app/ami.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+                ami = Ami.build_model(region_name, ami_dict)
+                amis.append(ami)
+        return amis
+
+    # Get the info about a single AMI
+    @staticmethod
+    def build_model(region_name: str, resource_dict: dict):
+        tags = resource.make_tags_dict(resource_dict.get('Tags', []))
+
+        state = resource_dict['State']
```
Comment (CodyRay) [inline app/ami.py:80]: What kind of states does an AMI have?

File: app/instance.py [hunk 5]
```diff
@@ -24,7 +23,7 @@ class Instance(Resource):
     # Return the type and state of the EC2 resource being examined ('instance' and 'running')
     @staticmethod
     def to_string():
-        return 'instance', 'running'
+        return 'instance', 'running'  # TODO: why do we assume this is running?
```
Comment (CodyRay) [inline app/instance.py]: Good question, I thought this was only used for building messages?

File: app/nagbot.py [hunk 6]
```diff
@@ -33,63 +35,59 @@
...
     def notify_internal(channel, dryrun):
-        summary_msg = "Hi, I'm Nagbot v{} :wink: ".format(__version__)
+        summary_msg = f"Hi, I'm Nagbot v{__version__} :wink: "
         summary_msg += "My job is to make sure we don't forget about unwanted AWS resources and waste money!\n"
 
         for resource_type in RESOURCE_TYPES:
             ec2_type, ec2_state = resource_type.to_string()
             resources = resource_type.list_resources()
-            num_running_resources = sum(1 for r in resources if r.state == 'running' or r.state == 'available')
+            num_running_resources = sum(1 for r in resources if r.state in ['running', 'available', 'completed'])
```
Comment (CodyRay) [inline app/nagbot.py]: (refactoring opportunity) resources could have a `is_running` method that handles this to avoid the state leaking out like this

File: app/nagbot.py [hunk 7]
```diff
@@ -33,63 +35,59 @@
...
 
             resources = sorted((r for r in resources if not (len(r.eks_nodegroup_name) > 0)), key=lambda i: i.name)
 
             resources_to_terminate = (list(r for r in resources if r.is_terminatable(TODAY_YYYY_MM_DD)))
             resources_to_stop = list(r for r in resources if r.is_stoppable(today_date=TODAY_YYYY_MM_DD))
 
             if len(resources_to_terminate) > 0:
-                summary_msg += 'The following %d _stopped_ {}s are due to be *TERMINATED*, ' \
-                                'based on the "Terminate after" tag:\n'.format(ec2_type) \
-                                % len(resources_to_terminate)
```
Comment (CodyRay) [inline app/nagbot.py:60]: Oof, this one had both `.format` and `%` based formatting - nice to see strings cleaned up :+1:

File: app/nagbot.py [hunk 8]
```diff
@@ -33,63 +35,59 @@
...
+                else:
+                    summary_msg += f'No {ec2_type}s are due to be stopped at this time.\n'
         sqslack.send_message(channel, summary_msg)
 
     def notify(self, channel, dryrun):
         try:
             self.notify_internal(channel, dryrun)
         except Exception as e:
-            sqslack.send_message(channel, "Nagbot failed to run the 'notify' command: " + str(e))
+            sqslack.send_message(channel, f"Nagbot failed to run the 'notify' command: {str(e)}")
```
Comment (CodyRay) [inline app/nagbot.py]: (nit) `str()` is redundant here now

File: app/nagbot.py [hunk 9]
```diff
@@ -108,37 +106,36 @@ def execute_internal(channel, dryrun):
...
-                        .format(r.stop_after, money_to_string(r.monthly_price), contact)
-                    resource.stop_resource(r.region_name, r.resource_id, dryrun=dryrun)
-                    resource.set_tag(r.region_name, r.ec2_type, r.resource_id, r.nagbot_state_tag_name, 'Stopped on '
-                                     + TODAY_YYYY_MM_DD, dryrun=dryrun)
-                sqslack.send_message(channel, message)
-            else:
-                sqslack.send_message(channel, 'No {}s were stopped today.'.format(ec2_type))
+                sqslack.send_message(channel, f'No {ec2_type}s were terminated today.')
+
+            if resource_type == "Instance":
```
Comment (CodyRay) [inline app/nagbot.py:120]: Why is this needed? Shouldn't snapshots and amis use the same method as volumes to skip this part?

File: app/resource.py [hunk 10]
```diff
@@ -74,6 +74,28 @@ def estimate_monthly_ebs_storage_price(region_name: str, instance_id: str, volum
         return size * 0.1
 
 
+# Estimated monthly costs were formulated by taking the average monthly costs of N. California and Oregon
+def estimate_monthly_snapshot_price(type: str, size: float) -> float:
```
Comment (CodyRay) [inline app/resource.py]: (refactoring opportunity) IMO many (all?) of the methods in this file could belong in their corresponding file (instance.py, etc)

File: app/resource.py [hunk 11]
```diff
@@ -74,6 +74,28 @@ def estimate_monthly_ebs_storage_price(region_name: str, instance_id: str, volum
...
+    standard_monthly_cost = .0525
+    archive_monthly_cost = .0131
+    return standard_monthly_cost*size if type == "standard" else archive_monthly_cost*size
+
+def estimate_monthly_ami_price(type: str, block_device_mappings: list, ami: str) -> float:
+    total_cost = 0
+    # Logic is only implemented for ebs-backed AMIs since Seeq does not use instance-backed AMIs
+    if type == 'ebs':
```
Comment (CodyRay) [inline app/resource.py]: I'd vote to log a warning if we get a type that isn't ebs

File: app/ami.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+    def is_stoppable_without_warning(self):
+        return self.generic_is_stoppable_without_warning(self)
+
+    # Check if an ami is stoppable (should always be false)
+    def is_stoppable(self, today_date, is_weekend=TODAY_IS_WEEKEND):
+        return self.generic_is_stoppable(self, today_date, is_weekend)
+
+    # Check if an ami is deletable/terminatable
+    def is_terminatable(self, today_date):
+        state = 'available'
```
Comment (CodyRay) [inline app/ami.py:152]: Is it possible to determine if the snapshot is associated with an AMI or AWS backup here to prevent us from trying to delete the snapshot?

File: app/ami.py [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+        return line
+
+    # Create ami url
+    @staticmethod
+    def url_from_id(region_name, resource_id):
+        resource_type = 'Amis'
+        return Resource.generic_url_from_id(region_name, resource_id, resource_type)
+
+    # Include ami in monthly price calculation if available
+    def included_in_monthly_price(self):
```
Comment (CodyRay) [inline app/ami.py:188]: I think currently the snapshots associated with amis will get double counted (counted for snapshot total and ami total). I'd vote to count the associated ones with the AMI total. Which I think you could do here by returning false if the snapshot is associated with an ami

**Review by CodyRay — APPROVED** [review]
Looking good, mostly learning questions from me

> I did not implement logic to explicitly delete the snapshots comprising EBS-backed AMIs since ideally those snapshots would have the same terminate date as the AMI upon creation.

I'd vote to have the AMI resource take care of deleting the snapshot, that feels more consistant with how we handled the similar relationship between instances and volumes. (though in that case AWS actually deletes the volumes from the instance for us)

**Comment by alizbay94** [comment]
> I'd vote to have the AMI resource take care of deleting the snapshot, that feels more consistant with how we handled the similar relationship between instances and volumes. (though in that case AWS actually deletes the volumes from the instance for us)

Sounds good i'll implement that so they are all deleted as one happy family.

File: app/ami.py (hunk 3 above)
Comment (alizbay94) [reply app/ami.py:53]: storage and iops are only available on some amis, but I just implemented a function that assigns iops and volume_type if it exists in the block_device_mappings, otherwise iops and volume_type will just be None.

File: README.md (hunk 1 above)
Comment (alizbay94) [reply README.md]: done

File: app/ami.py (hunk 4 above)
Comment (alizbay94) [reply app/ami.py:80]: I believe the only states are available and pending

File: app/instance.py (hunk 5 above)
Comment (alizbay94) [reply app/instance.py]: I think it is only for building messages, so I left it as is even though an AMI doesn't necessarily run.

File: app/ami.py (hunk 2 above)
Comment (alizbay94) [reply app/ami.py:10]: I don't think any changes would be necessary since that PR was only affecting `resource.py`, @sejal-raval does this sound right?

File: app/nagbot.py (hunk 9 above)
Comment (alizbay94) [reply app/nagbot.py:120]: I implemented this because the output on slack includes "No volumes are due to be stopped at this time" and since Volumes, Snapshots, and AMIs don't stop I thought it would look cleaner on slack to only display _stopped_ Instances.

File: app/resource.py (hunk 10 above)
Comment (alizbay94) [reply app/resource.py]: I agree, i'll make those changes. It just doesn't feel right putting those methods in the resource file 😄

File: app/ami.py (hunk 2 above)
Comment (sejal-raval) [reply app/ami.py:10]: That's a good point. https://github.com/seeq12/nagbot/pull/11 changed that all generic methods in `resource.py` reads time in Pacific TimeZone. I think it should be changed for the individual resources e.g. instances, volumes etc. for concurrency. I can open another PR for that.

File: app/ami.py (hunk 12 above)
Comment (alizbay94) [reply app/ami.py:152]: Yes, that information is included in the "Description" of the dictionary returned by the aws call - I just added two additional attributes in the Snapshot object, `is_ami_snapshot` and `is_aws_backup` snapshot. If the snapshot is either it will not get deleted, and I adjusted the pytests to test for that scenario.

File: app/ami.py (hunk 13 above)
Comment (alizbay94) [reply app/ami.py:188]: Done, I made it so that if it `is_ami_snapshot` its not included in the monthly price

File: app/nagbot.py (hunk 6 above)
Comment (alizbay94) [reply app/nagbot.py]: Done, I added a method called `is_active` to each of the resources

File: app/resource.py (hunk 11 above)
Comment (alizbay94) [reply app/resource.py]: Done I used a print statement with details of the AMI if it is not ebs-backed

File: app/resource.py (hunk 10 above)
Comment (alizbay94) [reply app/resource.py]: Done, moved the estimate monthly cost functions from resource.py to their respective files

File: app/nagbot.py (hunk 9 above)
Comment (CodyRay) [reply app/nagbot.py:120]: That seems like a good idea. Maybe there is a way to define if a resource type is "stopable" generically?

File: app/ami.py [hunk 14]
```diff
@@ -143,6 +144,10 @@ def is_safe_to_terminate(self, today_date):
         resource_type = Ami
         return self.generic_is_safe_to_terminate(self, resource_type, today_date)
 
+    # Check if a instance is active
+    def is_active(self):
+        return True if self.state == 'available' else False
```
Comment (CodyRay) [inline app/ami.py]: Can be simplified to 

```suggestion
        return self.state == 'available'
```

File: app/ami.py [hunk 15]
```diff
@@ -176,3 +181,22 @@ def get_ami_iops_and_volume_type(block_device_mappings):
...
+            # but we do not contain any s3 with bundles for AMIs, so these ephemeral volumes should only cost money
+            # when an instance is fired up from the AMI, therefore this cost is not included in the sum total.
+            if "Ebs" in device.keys():
+                snapshot = device["Ebs"]
+                snapshot_type = snapshot["VolumeType"]
+                snapshot_size = snapshot["VolumeSize"]
+                total_cost += estimate_monthly_snapshot_price(snapshot_type, snapshot_size)
+    else:
+        print(f"WARNING: {ami_name} is a {ami_type} type AMI with the following block_device_mappings: "
+              "{block_device_mappings}")
```
Comment (CodyRay) [inline app/ami.py]: Looks like an `f` was lost
```suggestion
              f"{block_device_mappings}")
```

File: app/snapshot.py [hunk 16]
```diff
@@ -77,30 +83,39 @@ def build_model(region_name: str, resource_dict: dict):
         snapshot_type = resource_dict['StorageTier']
         resource_id_tag = 'SnapshotId'
         resource_type_tag = 'StorageTier'
-        monthly_price = resource.estimate_monthly_snapshot_price(snapshot_type, size)
+        is_aws_backup_snapshot = False
+        is_ami_snapshot = False
+        if "AWS Backup service" in resource_dict['Description']:
+            is_aws_backup_snapshot = True
+        elif "Copied for DestinationAmi" in resource_dict['Description']:
```
Comment (CodyRay) [inline app/snapshot.py]: This might not be the most rebust method for Amis in particular because if someone deletes an AMI manually and doesn't delete the snapshot, then nagbot won't ever clean it up.

**Review by CodyRay — APPROVED** [review]

File: app/snapshot.py (hunk 16 above)
Comment (alizbay94) [reply app/snapshot.py]: Good point. I could extract the AMI's ID from the description of the snapshot, check if it exists with boto3, if it doesn't then the snapshot could be considered terminatable / part of the overall cost.

File: app/nagbot.py (hunk 9 above)
Comment (alizbay94) [reply app/nagbot.py:120]: I see, so it would be

`if resource_type.can_be_stopped()`

rather than 

`if resource_type == "Instance"`

I can implement that

File: app/snapshot.py (hunk 16 above)
Comment (alizbay94) [reply app/snapshot.py]: Implemented logic for the following:
If snapshot is an ami snapshot, we use a regex to match the first occurrence of an ami id since the first occurrence refers to the 'Destination AMI' (what we want) instead of the 'Source AMI'.
That ami id is checked against all the orgs amis to see if it is registered - if it is not registered the snapshot is not considered an ami snapshot and it will be treated like the rest of its bretheren able to be deleted and included in the total snapshot cost. (After running this looks like we have a good amount of ami snapshots whose amis have been deregistered so we should save a nice chunk of change by deleting them!)
One note - I used a global variable for the orgs amis so that the API call would only be made rather than once for every ami snapshot since the call is costly.

File: app/nagbot.py (hunk 9 above)
Comment (alizbay94) [reply app/nagbot.py:120]: done

File: app/ami.py [hunk 17]
```diff
@@ -198,5 +204,15 @@ def estimate_monthly_ami_price(ami_type: str, block_device_mappings: list, ami_n
...
+
+
+def is_ami_registered(ami_id: str, ) -> bool:
+    # Use global variable to make API call once instead of every time the method is called since call is costly
+    global ORG_AMIS
+    if not ORG_AMIS:
+        ec2_client = boto3.client('ec2')
+        ORG_AMIS = ec2_client.describe_images(Owners=['self'])
+    # Iterate through each image owned by organization to check if ami id exists, if it does it is registered
+    return True if [i['ImageId'] for i in ORG_AMIS['Images'] if i['ImageId'] == ami_id] else False
```
Comment (CodyRay) [inline app/ami.py]: Instead of loading a list of images, I think you could request just `ami_id` to load a single image. If it succeeds, then it must be registered. If boto3 fails to load it then it probably doesn't exist. I think that would be efficient enough that a global variable wouldn't be needed

**Review by CodyRay — APPROVED** [review]

File: app/ami.py (hunk 17 above)
Comment (alizbay94) [reply app/ami.py]: Done.

I ended up using 2 exceptions for the try/except  attempting to find a single AMI with boto3 because typically, if an ami does not exist, `ec2.Image(ami_id).name` throws a `botocore.exceptions.ClientError`, but I encountered a case where an `AttributeError` is thrown that reads: `AttributeError: 'NoneType' object has no attribute 'get'`

After doing a google search, I discovered a bug with an open ticket [here](https://github.com/boto/boto3/issues/925) on the boto3 project’s GitHub page describing how None type is returned for deregistered images, so I included the `AttributeError` as well to cover those cases.

File: app/ami.py [hunk 18]
```diff
@@ -209,10 +209,11 @@ def estimate_monthly_ami_price(ami_type: str, block_device_mappings: list, ami_n
...
-        ec2_client = boto3.client('ec2')
-        ORG_AMIS = ec2_client.describe_images(Owners=['self'])
-    # Iterate through each image owned by organization to check if ami id exists, if it does it is registered
-    return True if [i['ImageId'] for i in ORG_AMIS['Images'] if i['ImageId'] == ami_id] else False
+    ec2 = boto3.resource('ec2')
+    is_registered = True
+    # Retrieve name of AMI, if ClientError or AttributeError is thrown, the AMI does not exist
+    try:
+        ec2.Image(ami_id).name
+    except (botocore.exceptions.ClientError, AttributeError):
```
Comment (CodyRay) [inline app/ami.py:233]: `AttributeError` seems suspicious...

File: app/ami.py [hunk 19]
```diff
@@ -209,10 +209,11 @@ def estimate_monthly_ami_price(ami_type: str, block_device_mappings: list, ami_n
 
 
 def is_ami_registered(ami_id: str, ) -> bool:
```
Comment (CodyRay) [inline app/ami.py]: Seems to be an extra `,`
```suggestion
def is_ami_registered(ami_id: str) -> bool:
```

File: requirements.txt [hunk 20]
```diff
@@ -1,4 +1,5 @@
 boto3==1.20.26
+botocore==1.23.54
```
Comment (CodyRay) [inline requirements.txt:2]: Should these two be the same version?

**Review by CodyRay — APPROVED** [review]

File: app/ami.py (hunk 18 above)
Comment (alizbay94) [reply app/ami.py:233]: I agree - I don't know if you saw my previous comment, but I discovered a bug with an open ticket [here](https://github.com/boto/boto3/issues/925) on the boto3 project’s GitHub page describing how None type is returned for deregistered amis, and the error that is thrown is: `AttributeError: 'NoneType' object has no attribute 'get'`.
If its too suspicious I can always go back to the other method of getting a list of our AMIs and checking them. I'm alright with using the `AttributeError` with the `ClientError` because I can't anticipate serious ramifications, but I may be overlooking something.

File: requirements.txt (hunk 20 above)
Comment (alizbay94) [reply requirements.txt:2]: They should work as different versions, I looked online and saw the [latest version of boto3](https://github.com/boto/boto3/tags) is 1.25.3 and [the latest version of botocore](https://github.com/boto/botocore/tags) is 1.28.3. I imagine if they had to be the same version the latest versions would be the same, and there weren't any conflicts on my machine when I ran nagbot (I have those two versions running).

**Review by CodyRay — APPROVED** [review]

**Comment by CodyRay** [comment]
> > I'd vote to have the AMI resource take care of deleting the snapshot, that feels more consistant with how we handled the similar relationship between instances and volumes. (though in that case AWS actually deletes the volumes from the instance for us)
> 
> Sounds good i'll implement that so they are all deleted as one happy family.

Wanted to give this a poke because I think it is the only piece of feedback from me AFAIK

**Comment by alizbay94** [comment]
Ended up extracting the snapshot ids from the block device mappings of AMIs, then created aws ec2.Snapshot objects using those extracted ids. After the list of snapshot objects is created, it is assigned to the "snapshots" attribute of the Ami object that is created. I added a "snapshots" attribute to the Ami class that represents a list of aws ec2.Snapshot objects that make up the Ami. During the method when an Ami is terminated (deregistered), each of the Snapshots belonging to the Ami are also deleted.

**Review by CodyRay — APPROVED** [review]

**Review by CodyRay — APPROVED** [review]
